File: JavaVersion/src/GraphingPackage/Vertex.java
package GraphingPackage;

import java.util.ArrayList;

public class Vertex
{
   private String name;
   private int degree;
   private int vertexNumber;
   private ArrayList<Edge> edgeSet;
   
   private static char defaultName = 'A';
   private static int numVertices = 0;
   

   /**
    * Creates a new vertex
    * @param name The name of the Vertex
    */
   public Vertex ( String name )
   {
      this.name = name;
      degree = 0;
      edgeSet = new ArrayList<>();
      setVertexNumber();
   }
   
   public Vertex () {
      name = "" + defaultName;
      defaultName++;
      degree = 0;
      edgeSet = new ArrayList<>();
      setVertexNumber();
   }
   
   /**
    * Sets the vertex number and increments the amount of vertices
    */
   private void setVertexNumber()
   {
      vertexNumber = numVertices;
      numVertices++;
   }
   
   /**
    * Returns the vertex number
    * @return the vertex number
    */
   public int getVertexNumber()
   {
      return vertexNumber;
   }
   
   /**
    * Gets the name of the vertex
    * @return the name of the vertex
    */
   public String getName()
   {
      return name;
   }
   
   /**
    * Returns the number of vertices
    * @return the number of vertices
    */
   public int getNumvertices()
   {
      return numVertices;
   }
   
   /**
    * Returns the degree of the Vertex
    * @return the degree of the vertex
    */
   public int getDegree()
   {
      return degree;
   }
   
   /**(
    * Increases the degree of the vertex
    */
   public void increaseDegree()
   {
      degree++;
   }
   
   /**
    * Returns the shortest neighbor from this vertex
    * @param lowestAcceptableEdge The amount of times the lowest edge has been searched for
    * @return The shortest edge
    */
   public Edge getShortestNeighbor( int lowestAcceptableEdge )
   {
      sortEdges();
      return edgeSet.get( lowestAcceptableEdge );
   }
   
   /**
    * Adds an Edge to the edge list
    * @param newEdge The edge to be added
    */
   public void addEdge( Edge newEdge )
   {
      edgeSet.add( newEdge );
   }
   
   /**
    * Sorts the edges from least weight to the most using selection sort
    */
   public void sortEdges()
   {
      int listIndex, innerIndex, smallestWeight, innerComparison, smallestWeightIndex;
      for ( listIndex = 0; listIndex < edgeSet.size() - 1; listIndex++ )
      {
         smallestWeightIndex = listIndex;
         for ( innerIndex = listIndex + 1; innerIndex < edgeSet.size(); innerIndex++ )
         {
            smallestWeight = edgeSet.get( smallestWeightIndex ).getWeight();
            innerComparison = edgeSet.get( innerIndex ).getWeight();
            if ( innerComparison < smallestWeight )
            {
               smallestWeightIndex = innerIndex;
            }
         }
         swapElements( listIndex, smallestWeightIndex );
      }
   }
   

   private void swapElements( int oneIndex, int otherIndex )
   {
      Edge temp = edgeSet.get( oneIndex );
      
      edgeSet.set( oneIndex, edgeSet.get( otherIndex ) );
      
      edgeSet.set( otherIndex, temp );
   }
   
   /**
    * Prints out all the edges of the vertex
    * <p>
    * Note: used for testing purposes
    */
   public void displayEdges()
   {
      int index;
      for ( index = 0; index < edgeSet.size(); index++ )
      {
         System.out.println( edgeSet.get(index).toString() );
      }
   }
   
   /**
    * Dumps the name of the Vertex
    */
   public String toString()
   {
      return "The name of the vertex is: " + name;
   }
   
   public boolean equals( Vertex otherVertex )
   {
      return vertexNumber == otherVertex.vertexNumber;
   }
}

File: PythonVersion/circuits/vertex.py
class Vertex:
    _default_name = "A"
    _num_vertices = 0

    def __init__( self, name=None ):
        if name is None:
            self._name = Vertex._default_name
            Vertex._default_name = chr( ord( Vertex._default_name[0] ) + 1 )
        else:
            self._name = name
        self._degree = 0
        self._edge_set = []
        self._vertex_number = Vertex._num_vertices
        Vertex._num_vertices += 1

    def get_vertex_number(self):
        return self._vertex_number

    def get_name(self):
        return self._name

    def get_num_vertices(self):
        return self._num_vertices

    def get_degree(self):
        return self._degree

    def increase_degree( self ):
        self._degree += 1

    def get_shortest_neighbor(self, lowest_acceptable_edge ):
        self.sort_edges()
        return self._edge_set[ lowest_acceptable_edge ]

    def add_edge( self, new_edge ):
        self._edge_set.append( new_edge )

    def sort_edges(self):
        list_index = 0
        smallest_weight = 0
        smallest_weight_index = 0
        inner_comparison = 0
        while list_index < len( self._edge_set ):
            smallest_weight_index = list_index
            inner_index = list_index + 1
            while inner_index < len( self._edge_set ):
                smallest_weight = self._edge_set[ smallest_weight_index ].get_weight()
                inner_comparison = self._edge_set[ inner_index ].get_weight()
                if( inner_comparison < smallest_weight ):
                    smallest_weight_index = inner_index
                inner_index += 1
            self.swap_elements( list_index, smallest_weight_index )
            list_index += 1

    def swap_elements( self, one_index, other_index ):
        temp = self._edge_set[ one_index ]
        self._edge_set[ one_index ] = self._edge_set[ other_index ]
        self._edge_set[ other_index ] = temp

    def display_edges(self):
        index = 0
        while index < len( self._edge_set ):
            self._edge_set[ index ].to_string()
            index += 1

    def to_string(self):
        return  "The name of the vertex is: " + self._name

    def equals( self, other_vertex ):
        if self._vertex_number == other_vertex.get_vertex_number():
            return True
        return False

    
